test/e2e/upgrades/storage/volume_mode.go

```diff
@@ -0,0 +1,127 @@
+// formatted and mounted like a nil/Filesystem PV after a downgrade to a version
+// where the BlockVolume feature is disabled
+type VolumeModeDowngradeTest struct {
+	pvSource *v1.PersistentVolumeSource
+	pv       *v1.PersistentVolume
+	pvc      *v1.PersistentVolumeClaim
+	pod      *v1.Pod
+}
+
+func (VolumeModeDowngradeTest) Name() string {
+	return "[sig-storage] volume-mode-downgrade"
+}
+
+func (t *VolumeModeDowngradeTest) Skip(upgCtx upgrades.UpgradeContext) bool {
+	if !framework.ProviderIs("gce", "gke") {
+		return true
+	}
+
+	// Only run when downgrading from >= 1.13 to < 1.13
+	blockVersion := version.MustParseSemantic("1.13.0")
+	if upgCtx.Versions[0].Version.LessThan(blockVersion) {
+		return true
+	}
+	if !upgCtx.Versions[1].Version.LessThan(blockVersion) {
+		return true
+	}
+
+	return false
+}
+
+// Setup creates a block pv and then verifies that a pod can consume it.  The pod writes data to the volume.
+func (t *VolumeModeDowngradeTest) Setup(f *framework.Framework) {
+
+	var err error
+
+	cs := f.ClientSet
+	ns := f.Namespace.Name
+
+	By("Initializing PV source")
+	t.pvSource, _ = framework.CreateGCEVolume()
+	block := v1.PersistentVolumeBlock
+	pvConfig := framework.PersistentVolumeConfig{
+		NamePrefix: "volume-mode-upgrade",
+		PVSource:   *t.pvSource,
+		VolumeMode: &block,
+	}
+	emptyStorageClass := ""
+	pvcConfig := framework.PersistentVolumeClaimConfig{
+		StorageClassName: &emptyStorageClass,
+		VolumeMode:       &block,
+	}
+
+	By("Creating the PV and PVC")
+	t.pv, t.pvc, err = framework.CreatePVPVC(f.ClientSet, pvConfig, pvcConfig, ns, true)
+	Expect(err).NotTo(HaveOccurred())
+	framework.ExpectNoError(framework.WaitOnPVandPVC(f.ClientSet, ns, t.pv, t.pvc))
+
+	By("Consuming the PV before downgrade")
+	t.pod, err = framework.CreateSecPod(cs, ns, []*v1.PersistentVolumeClaim{t.pvc}, false, "", false, false, framework.SELinuxLabel, nil, framework.PodStartTimeout)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Checking if persistent volume exists as expected volume mode")
+	utils.CheckVolumeModeOfPath(t.pod, block, devicePath)
+
+	By("Checking if read/write to persistent volume works properly")
+	utils.CheckReadWriteToPath(t.pod, block, devicePath)
+}
+
+// Test waits for the downgrade to complete, and then verifies that a pod can no
+// longer consume the pv as it is not mapped nor mounted into the pod
+func (t *VolumeModeDowngradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+	By("Waiting for downgrade to finish")
+	<-done
+
+	By("Verifying that nothing exists at the device path in the pod")
+	utils.VerifyExecInPodFail(t.pod, fmt.Sprintf("test -e %s", devicePath), 1)
+
+	By("Deleting the pod")
+	framework.ExpectNoError(framework.DeletePodWithWait(f, f.ClientSet, t.pod))
+}
+
+// Teardown cleans up any remaining resources.
+func (t *VolumeModeDowngradeTest) Teardown(f *framework.Framework) {
+	errs := framework.PVPVCCleanup(f.ClientSet, f.Namespace.Name, t.pv, t.pvc)
```

Comment:
Can the test validate that the objects was successfully deleted? (ie it disappears)

Reply:
yes, Teardown rewritten